implementations/rust/ockam/ockam_identity/src/identities/storage/sqlite_storage.rs

```diff
@@ -0,0 +1,185 @@
+use ockam_core::async_trait;
+use ockam_core::compat::sync::{Arc, Mutex};
+use ockam_core::compat::vec::Vec;
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::{Error, Result};
+use ockam_node::tokio::task::{self, JoinError};
+use rusqlite::{params, Connection};
+use std::fmt;
+use std::path::Path;
+use tokio_retry::strategy::{jitter, FixedInterval};
+use tokio_retry::Retry;
+use tracing::debug;
+
+/// Storage using the Sqlite database
+#[derive(Clone)]
+pub struct SqliteStorage {
+    /// Sqlite Connection
+    pub conn: Arc<Mutex<Connection>>,
+}
+
+impl fmt::Debug for SqliteStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("SqliteStore")
+    }
+}
+
+impl SqliteStorage {
+    const CREATE_IDENTITY_TABLE_SQL: &str = "CREATE TABLE IF NOT EXISTS identity (
+        id INTEGER PRIMARY KEY,
+        identity_id TEXT NOT NULL,
+        key TEXT NOT NULL,
+        value BLOB
+    );";
+    const CREATE_IDENTITY_INDEX_SQL: &str =
+        "CREATE UNIQUE INDEX IF NOT EXISTS idx_identity_id_key ON identity (identity_id, key);";
+
+    const CREATE_POLICY_TABLE_SQL: &str = "CREATE TABLE IF NOT EXISTS policy (
+        id INTEGER PRIMARY KEY,
+        resource TEXT NOT NULL,
+        action TEXT NOT NULL,
+        value BLOB
+    );";
+    const CREATE_POLICY_INDEX_SQL: &str = "CREATE UNIQUE INDEX IF NOT EXISTS idx_policy_resource_action ON policy (resource, action);";
+
+    /// Constructor
+    pub async fn new<P: AsRef<Path>>(p: P) -> Result<Self> {
+        // Not sure we need this
+        // creating a new database might be failing a few times
+        // if the files are currently being held by another pod which is shutting down.
+        // In that case we retry a few times, between 1 and 10 seconds.
+        let retry_strategy = FixedInterval::from_millis(1000)
+            .map(jitter) // add jitter to delays
+            .take(10); // limit to 10 retries
+
+        let path: &Path = p.as_ref();
+        Retry::spawn(retry_strategy, || async { Self::make(path).await }).await
+    }
+
+    async fn make(p: &Path) -> Result<Self> {
+        debug!("create the Sqlite database");
+        let p = p.to_path_buf();
+        // Creates database file if it doesn't exist
+        let conn = Connection::open(p).map_err(map_sqlite_err)?;
+        let _ = conn
+            .execute_batch(
+                &("PRAGMA encoding = 'UTF-8';".to_owned()
+                    + SqliteStorage::CREATE_IDENTITY_TABLE_SQL
+                    + SqliteStorage::CREATE_IDENTITY_INDEX_SQL
+                    + SqliteStorage::CREATE_POLICY_TABLE_SQL
+                    + SqliteStorage::CREATE_POLICY_INDEX_SQL),
+            )
+            .map_err(map_sqlite_err)?;
+        Ok(SqliteStorage {
+            conn: Arc::new(Mutex::new(conn)),
+        })
+    }
+}
+
+#[async_trait]
+impl Storage for SqliteStorage {
+    async fn get(&self, id: &str, key: &str) -> Result<Option<Vec<u8>>> {
+        let db = self.clone();
```

Comment:
Does it make sense to clone only the connection object?

Reply:
I don't think there will be any difference in how it actually works under the hood. I noticed that we use only the connection struct itself, so I think it makes sense to pass only the required object and nothing more. There is only one field in that struct, though, so there is not much difference.